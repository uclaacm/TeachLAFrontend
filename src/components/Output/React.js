import { getJsSrcDocLoggingScript } from "./constants";

const getReactSrcDocHead = () => `
    <head>
        <style>html,body: {margin:0, width:100%}</style>
        <script src="https://unpkg.com/@babel/standalone/babel.min.js"></script>
        <script src="https://unpkg.com/react@17/umd/react.development.js" crossorigin></script>
        <script src="https://unpkg.com/react-dom@17/umd/react-dom.development.js" crossorigin></script>
        <script src="https://unpkg.com/@material-ui/core@latest/umd/material-ui.development.js" crossorigin="anonymous"></script>
        <style>
          #inner {
            height:100px;
            background-color:#222;
            color: #DDD;
            font-family: monospace;
            word-wrap:break-word;
            overflow:auto;
            margin: 10px auto;
            position:relative;
            padding: 10px 35px 10px 10px;
            width: 100%;
            box-sizing: border-box;         /* For IE and modern versions of Chrome */
            -moz-box-sizing: border-box;    /* For Firefox                          */
            -webkit-box-sizing: border-box; /* For Safari                           */
          }
          #output {
            display: block;
            position: relative;
            background: white;
            min-height: 70vh;
            padding: 10px 35px 10px 10px;
          }
          #closeConsoleButton { position: fixed; top: 20px; right: 30px; color: #ddd;}
        </style>
    </head>
  `;

const getUserScript = (code) => `
  <script type="text/babel">
    ${code}
    ReactDOM.render(
      <App />,
      document.getElementById('output'),
    );
  </script>
`;

const getReactSrcDocBody = (code, showConsole) => `
    <body>
      ${
        showConsole
          ? `<div id="outer"><textarea id="inner"></textarea></div>`
          : `<div id="outer" style="display:none;"><textarea id="inner"></textarea></div>`
      }
      ${getJsSrcDocLoggingScript()}
      ${getUserScript(code)}
      <div id="output"></div>
    </body>
  `;

export default function (code, showConsole) {
  return `<html> ${getReactSrcDocHead()} ${getReactSrcDocBody(code, showConsole)}</html>`;
}
