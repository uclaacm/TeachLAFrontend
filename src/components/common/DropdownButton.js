import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import React, { useState } from 'react';
import {
  Dropdown, DropdownToggle, DropdownMenu, DropdownItem,
} from 'reactstrap';

/** --------Props---------------
 * dropDownItems: array of data for each dropdown item
 * displayValue: string to be displayed as the placeholder for the dropdown
 * toggleProps: props that dictate the styling of dropDownToggle
 * icon: contains optional icon for editor dropDown
 * displayClass: contains different classes for editor and sketch dropdownbutton
 */
/** --------Optional props--------
 * defaultOpen: boolean determining if the dropdown should start off open or closed
 */

const DropdownButton = function (props) {
  const {
    uid,
    icon,
    DropdownItems,
    defaultOpen,
    displayValue,
    displayClass,
    toggleProps,
    onSelect,
    dirty,
  } = props;

  const dropDownParentClass = `${displayClass}-language-dropdown`;
  const dropDownItemClass = `${displayClass}-language-dropdown-closed-content`;

  const [dropdownOpen, setdropdownOpen] = useState(defaultOpen || false);

  const toggleHandler = () => {
    setdropdownOpen(!dropdownOpen);
  };

  const renderDropdownItems = () => DropdownItems.map(({ display, value, icon }) => (
    <DropdownItem key={value} onClick={() => onSelect({
      display,
      value,
      dirty,
      uid,
    })}>
      <FontAwesomeIcon style={{ marginRight: '10px' }} icon={icon} fixedWidth />
      {display}
    </DropdownItem>
  ));

  return (
    <div className={dropDownParentClass}>
      <Dropdown isOpen={dropdownOpen} toggle={() => toggleHandler(dropdownOpen)}>
        <DropdownToggle caret {...toggleProps}>
          <div className={dropDownItemClass}>
            <FontAwesomeIcon icon={icon} fixedWidth />
            {displayValue}
          </div>
        </DropdownToggle>
        <DropdownMenu>{renderDropdownItems()}</DropdownMenu>
      </Dropdown>
    </div>
  );
};
export default DropdownButton;
