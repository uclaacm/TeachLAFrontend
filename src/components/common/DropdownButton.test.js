import { shallow } from 'enzyme';
import React from 'react';
import { DropdownItem } from 'reactstrap';
import DropdownButton from './DropdownButton';

const validDropdownItems = [
  {
    display: 'Nice',
    value: 'banana',
    icon: 'cubes',
  },
  {
    display: 'Cool',
    value: 'apple',
    icon: 'mobile',
  },
  {
    display: 'Fun',
    value: 'pear',
    icon: 'wind',
  },
];

describe('DropdownButton', () => {
  it('smoke test', () => {
    const component = shallow(<DropdownButton DropdownItems={validDropdownItems} />);
    expect(component.exists()).toBe(true);
  });

  it('renders every dropdown item correctly', () => {
    const component = shallow(
      <DropdownButton
        icon="bacteria"
        DropdownItems={validDropdownItems}
        defaultOpen
        displayValue="bar"
      />,
    );
    expect(component).toMatchSnapshot();
  });

  it('handles defaultOpen properly', () => {
    const componentTrue = shallow(
      <DropdownButton
        icon="lungs"
        DropdownItems={validDropdownItems}
        defaultOpen
        displayValue="foo"
      />,
    );
    expect(componentTrue).toMatchSnapshot();

    const componentDefault = shallow(
      <DropdownButton icon="lungs" DropdownItems={validDropdownItems} displayValue="foo" />,
    );
    expect(componentDefault).toMatchSnapshot();

    const componentFalse = shallow(
      <DropdownButton
        icon="lungs"
        DropdownItems={validDropdownItems}
        defaultOpen={false}
        displayValue="foo"
      />,
    );
    expect(componentFalse).toMatchSnapshot();
  });

  it('clicking an option triggers the handleClick function', () => {
    const clickFn = jest.fn(({ display, value, icon }) => ({ display, value, icon }));

    // check that clicking the selected value calls the handleClick fn

    const component = shallow(
      <DropdownButton
        icon="lungs"
        DropdownItems={validDropdownItems}
        defaultOpen
        displayValue="foo"
        displayClass="bar"
        onSelect={clickFn}
      />,
    );

    component.find(DropdownItem).at(0).simulate('click');
    expect(clickFn.mock.calls[0][0]).toStrictEqual({
      dirty: undefined,
      display: 'Nice',
      value: 'banana',
    });

    const component2 = shallow(
      <DropdownButton
        icon="bacteria"
        DropdownItems={validDropdownItems}
        defaultOpen
        displayValue="foo"
        displayClass="bar"
        onSelect={clickFn}
      />,
    );

    component2.find(DropdownItem).at(1).simulate('click');
    expect(clickFn.mock.calls[1][0]).toStrictEqual({
      dirty: undefined,
      display: 'Cool',
      value: 'apple',
    });
  });
});
