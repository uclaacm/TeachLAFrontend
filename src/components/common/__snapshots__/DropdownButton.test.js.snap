// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownButton handles defaultOpen properly 1`] = `
<div
  className="undefined-language-dropdown"
>
  <Dropdown
    a11y={true}
    active={false}
    addonType={false}
    direction="down"
    inNavbar={false}
    isOpen={true}
    nav={false}
    setActiveFromChild={false}
    toggle={[Function]}
  >
    <DropdownToggle
      aria-haspopup={true}
      caret={true}
      color="secondary"
    >
      <div
        className="undefined-language-dropdown-closed-content"
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="lungs"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        foo
      </div>
    </DropdownToggle>
    <DropdownMenu
      flip={true}
      tag="div"
    >
      <DropdownItem
        key="banana"
        onClick={[Function]}
        tag="button"
        toggle={true}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="cubes"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          style={
            Object {
              "marginRight": "10px",
            }
          }
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        Nice
      </DropdownItem>
      <DropdownItem
        key="apple"
        onClick={[Function]}
        tag="button"
        toggle={true}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="mobile"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          style={
            Object {
              "marginRight": "10px",
            }
          }
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        Cool
      </DropdownItem>
      <DropdownItem
        key="pear"
        onClick={[Function]}
        tag="button"
        toggle={true}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="wind"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          style={
            Object {
              "marginRight": "10px",
            }
          }
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        Fun
      </DropdownItem>
    </DropdownMenu>
  </Dropdown>
</div>
`;

exports[`DropdownButton handles defaultOpen properly 2`] = `
<div
  className="undefined-language-dropdown"
>
  <Dropdown
    a11y={true}
    active={false}
    addonType={false}
    direction="down"
    inNavbar={false}
    isOpen={false}
    nav={false}
    setActiveFromChild={false}
    toggle={[Function]}
  >
    <DropdownToggle
      aria-haspopup={true}
      caret={true}
      color="secondary"
    >
      <div
        className="undefined-language-dropdown-closed-content"
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="lungs"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        foo
      </div>
    </DropdownToggle>
    <DropdownMenu
      flip={true}
      tag="div"
    >
      <DropdownItem
        key="banana"
        onClick={[Function]}
        tag="button"
        toggle={true}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="cubes"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          style={
            Object {
              "marginRight": "10px",
            }
          }
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        Nice
      </DropdownItem>
      <DropdownItem
        key="apple"
        onClick={[Function]}
        tag="button"
        toggle={true}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="mobile"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          style={
            Object {
              "marginRight": "10px",
            }
          }
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        Cool
      </DropdownItem>
      <DropdownItem
        key="pear"
        onClick={[Function]}
        tag="button"
        toggle={true}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="wind"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          style={
            Object {
              "marginRight": "10px",
            }
          }
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        Fun
      </DropdownItem>
    </DropdownMenu>
  </Dropdown>
</div>
`;

exports[`DropdownButton handles defaultOpen properly 3`] = `
<div
  className="undefined-language-dropdown"
>
  <Dropdown
    a11y={true}
    active={false}
    addonType={false}
    direction="down"
    inNavbar={false}
    isOpen={false}
    nav={false}
    setActiveFromChild={false}
    toggle={[Function]}
  >
    <DropdownToggle
      aria-haspopup={true}
      caret={true}
      color="secondary"
    >
      <div
        className="undefined-language-dropdown-closed-content"
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="lungs"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        foo
      </div>
    </DropdownToggle>
    <DropdownMenu
      flip={true}
      tag="div"
    >
      <DropdownItem
        key="banana"
        onClick={[Function]}
        tag="button"
        toggle={true}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="cubes"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          style={
            Object {
              "marginRight": "10px",
            }
          }
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        Nice
      </DropdownItem>
      <DropdownItem
        key="apple"
        onClick={[Function]}
        tag="button"
        toggle={true}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="mobile"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          style={
            Object {
              "marginRight": "10px",
            }
          }
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        Cool
      </DropdownItem>
      <DropdownItem
        key="pear"
        onClick={[Function]}
        tag="button"
        toggle={true}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="wind"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          style={
            Object {
              "marginRight": "10px",
            }
          }
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        Fun
      </DropdownItem>
    </DropdownMenu>
  </Dropdown>
</div>
`;

exports[`DropdownButton handles dropdown click correctly 1`] = `
<div
  className="undefined-language-dropdown"
>
  <Dropdown
    a11y={true}
    active={false}
    addonType={false}
    direction="down"
    inNavbar={false}
    isOpen={false}
    nav={false}
    setActiveFromChild={false}
    toggle={[Function]}
  >
    <DropdownToggle
      aria-haspopup={true}
      caret={true}
      color="secondary"
    >
      <div
        className="undefined-language-dropdown-closed-content"
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="lungs"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        foo
      </div>
    </DropdownToggle>
    <DropdownMenu
      flip={true}
      tag="div"
    >
      <DropdownItem
        key="banana"
        onClick={[Function]}
        tag="button"
        toggle={true}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="cubes"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          style={
            Object {
              "marginRight": "10px",
            }
          }
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        Nice
      </DropdownItem>
      <DropdownItem
        key="apple"
        onClick={[Function]}
        tag="button"
        toggle={true}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="mobile"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          style={
            Object {
              "marginRight": "10px",
            }
          }
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        Cool
      </DropdownItem>
      <DropdownItem
        key="pear"
        onClick={[Function]}
        tag="button"
        toggle={true}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="wind"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          style={
            Object {
              "marginRight": "10px",
            }
          }
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        Fun
      </DropdownItem>
    </DropdownMenu>
  </Dropdown>
</div>
`;

exports[`DropdownButton handles dropdown click correctly 2`] = `
<div
  className="undefined-language-dropdown"
>
  <Dropdown
    a11y={true}
    active={false}
    addonType={false}
    direction="down"
    inNavbar={false}
    isOpen={true}
    nav={false}
    setActiveFromChild={false}
    toggle={[Function]}
  >
    <DropdownToggle
      aria-haspopup={true}
      caret={true}
      color="secondary"
    >
      <div
        className="undefined-language-dropdown-closed-content"
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="lungs"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        foo
      </div>
    </DropdownToggle>
    <DropdownMenu
      flip={true}
      tag="div"
    >
      <DropdownItem
        key="banana"
        onClick={[Function]}
        tag="button"
        toggle={true}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="cubes"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          style={
            Object {
              "marginRight": "10px",
            }
          }
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        Nice
      </DropdownItem>
      <DropdownItem
        key="apple"
        onClick={[Function]}
        tag="button"
        toggle={true}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="mobile"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          style={
            Object {
              "marginRight": "10px",
            }
          }
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        Cool
      </DropdownItem>
      <DropdownItem
        key="pear"
        onClick={[Function]}
        tag="button"
        toggle={true}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="wind"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          style={
            Object {
              "marginRight": "10px",
            }
          }
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        Fun
      </DropdownItem>
    </DropdownMenu>
  </Dropdown>
</div>
`;

exports[`DropdownButton renders every dropdown item correctly 1`] = `
<div
  className="undefined-language-dropdown"
>
  <Dropdown
    a11y={true}
    active={false}
    addonType={false}
    direction="down"
    inNavbar={false}
    isOpen={true}
    nav={false}
    setActiveFromChild={false}
    toggle={[Function]}
  >
    <DropdownToggle
      aria-haspopup={true}
      caret={true}
      color="secondary"
    >
      <div
        className="undefined-language-dropdown-closed-content"
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="bacteria"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        bar
      </div>
    </DropdownToggle>
    <DropdownMenu
      flip={true}
      tag="div"
    >
      <DropdownItem
        key="banana"
        onClick={[Function]}
        tag="button"
        toggle={true}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="cubes"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          style={
            Object {
              "marginRight": "10px",
            }
          }
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        Nice
      </DropdownItem>
      <DropdownItem
        key="apple"
        onClick={[Function]}
        tag="button"
        toggle={true}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="mobile"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          style={
            Object {
              "marginRight": "10px",
            }
          }
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        Cool
      </DropdownItem>
      <DropdownItem
        key="pear"
        onClick={[Function]}
        tag="button"
        toggle={true}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={true}
          flip={null}
          icon="wind"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          style={
            Object {
              "marginRight": "10px",
            }
          }
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
        Fun
      </DropdownItem>
    </DropdownMenu>
  </Dropdown>
</div>
`;
